packages/service-registry-api/src/primitives.rs

```diff
@@ -0,0 +1,148 @@
+    pub bond_denom: String,
+    // should be set to a duration longer than the voting period for governance proposals,
+    // otherwise a verifier could bail before they get penalized
+    pub unbonding_period_days: u16,
+    pub description: String,
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, JsonSchema)]
+pub struct Verifier {
+    pub address: Addr,
+    pub bonding_state: BondingState,
+    pub authorization_state: AuthorizationState,
+    pub service_name: String,
+}
+
+impl Verifier {
+    pub fn bond(self, to_add: Option<nonempty::Uint128>) -> Result<Self, ContractError> {
+        let amount: nonempty::Uint128 = match self.bonding_state {
+            BondingState::Bonded { amount }
+            | BondingState::RequestedUnbonding { amount }
+            | BondingState::Unbonding {
+                amount,
+                unbonded_at: _,
+            } => amount
+                .into_inner()
+                .checked_add(to_add.map(Uint128::from).unwrap_or(Uint128::zero()))
+                .map_err(ContractError::Overflow)?
+                .try_into()?,
+            BondingState::Unbonded => to_add.ok_or(ContractError::NoFundsToBond)?,
+        };
+
+        Ok(Self {
+            bonding_state: BondingState::Bonded { amount },
+            ..self
+        })
+    }
+
+    pub fn unbond(self, can_unbond: bool, time: Timestamp) -> Result<Self, ContractError> {
+        if self.authorization_state == AuthorizationState::Jailed {
+            return Err(ContractError::VerifierJailed);
+        }
+
+        let bonding_state = match self.bonding_state {
+            BondingState::Bonded { amount } | BondingState::RequestedUnbonding { amount } => {
+                if can_unbond {
+                    BondingState::Unbonding {
+                        unbonded_at: time,
+                        amount,
+                    }
+                } else {
+                    BondingState::RequestedUnbonding { amount }
+                }
+            }
+            _ => return Err(ContractError::InvalidBondingState(self.bonding_state)),
+        };
+
+        Ok(Self {
+            bonding_state,
+            ..self
+        })
+    }
+
+    pub fn claim_stake(
+        self,
+        time: Timestamp,
+        unbonding_period_days: u64,
+    ) -> Result<(Self, nonempty::Uint128), ContractError> {
+        if self.authorization_state == AuthorizationState::Jailed {
+            return Err(ContractError::VerifierJailed);
+        }
+
+        match self.bonding_state {
+            BondingState::Unbonding {
+                amount,
+                unbonded_at,
+            } if unbonded_at.plus_days(unbonding_period_days) <= time => Ok((
+                Self {
+                    bonding_state: BondingState::Unbonded,
+                    ..self
+                },
+                amount,
+            )),
+            _ => Err(ContractError::InvalidBondingState(self.bonding_state)),
+        }
+    }
+}
```

Comment:
we'll have to look into how the verifier is used in other contracts. It doesn't make sense to provide this logic outside of the service-registry